lib/charms/layer/kubernetes_common.py

```diff
@@ -933,3 +934,199 @@ def get_secret_password(username):
+
+    # Put together the KubeletConfiguration data
+    kubelet_config = {
+        "apiVersion": "kubelet.config.k8s.io/v1beta1",
+        "kind": "KubeletConfiguration",
+        "address": "0.0.0.0",
+        "authentication": {
+            "anonymous": {"enabled": False},
+            "x509": {"clientCAFile": str(ca_crt_path)},
+        },
+        # NB: authz webhook config tells the kubelet to ask the api server
+        # if a request is authorized; it is not related to the authn
+        # webhook config of the k8s master services.
+        "authorization": {"mode": "Webhook"},
+        "clusterDomain": dns_domain,
+        "failSwapOn": False,
+        "port": 10250,
+        "protectKernelDefaults": True,
+        "readOnlyPort": 0,
+        "tlsCertFile": str(server_crt_path),
+        "tlsPrivateKeyFile": str(server_key_path),
+    }
+    if dns_ip:
+        kubelet_config["clusterDNS"] = [dns_ip]
+
+    # Handle feature gates
+    if get_version("kubelet") >= (1, 19):
+        # NB: required for CIS compliance
+        feature_gates["RotateKubeletServerCertificate"] = True
+    if is_state("kubernetes-worker.gpu.enabled"):
+        feature_gates["DevicePlugins"] = True
+    if feature_gates:
+        kubelet_config["featureGates"] = feature_gates
+    if is_dual_stack(cluster_cidr()):
+        feature_gates = kubelet_config.setdefault("featureGates", {})
+        feature_gates["IPv6DualStack"] = True
+
+    # Workaround for DNS on bionic
+    # https://github.com/juju-solutions/bundle-canonical-kubernetes/issues/655
+    resolv_path = os.path.realpath("/etc/resolv.conf")
+    if resolv_path == "/run/systemd/resolve/stub-resolv.conf":
+        kubelet_config["resolvConf"] = "/run/systemd/resolve/resolv.conf"
+
+    # Add kubelet-extra-config. This needs to happen last so that it
+    # overrides any config provided by the charm.
+    kubelet_extra_config = hookenv.config("kubelet-extra-config")
+    kubelet_extra_config = yaml.safe_load(kubelet_extra_config)
+    merge_kubelet_extra_config(kubelet_config, kubelet_extra_config)
+
+    # Render the file and configure Kubelet to use it
+    os.makedirs("/root/cdk/kubelet", exist_ok=True)
+    with open("/root/cdk/kubelet/config.yaml", "w") as f:
+        f.write("# Generated by kubernetes-worker charm, do not edit\n")
+        yaml.dump(kubelet_config, f)
+    kubelet_opts["config"] = "/root/cdk/kubelet/config.yaml"
+
+    # If present, ensure kubelet gets the pause container from the configured
+    # registry. When not present, kubelet uses a default image location
+    # (currently k8s.gcr.io/pause:3.4.1).
+    if registry:
+        kubelet_opts["pod-infra-container-image"] = get_sandbox_image_uri(registry)
+
+    if taints:
+        kubelet_opts["register-with-taints"] = ",".join(taints)
+
+    workaround_lxd_kernel_params()
+
+    configure_kubernetes_service(
+        "kubernetes-common.prev-args.", "kubelet", kubelet_opts, "kubelet-extra-args"
+    )
```

Comment:
Removed from [here](https://github.com/charmed-kubernetes/charm-kubernetes-worker/pull/103/files#diff-1837fd4ad476d5c87c628d68b3c461fc5c03d6aeafc928e910c69aa6ae267f85L760-L872) and added to this PR

the function signature changed from:
```python
def configure_kubelet(dns, ingress_ip):
   ...
```

to

```python
def configure_kubelet(dns_domain, dns_ip, registry, taints=None):
   ...
```

Reasons:
* `ingress_ip ` removed because it can be resolved from `get_node_ip()` in this module
* `dns` removed b/c it was a dictionary containing fields `'domain'` and `'sdn-ip'`  (not a typo)
  * `dns_domain` - takes over where `dns['domain']` was previously used
  * `dns_ip` - takes over where `dns['sdn-ip']` was previously used
* `registry` - added due to loss of scope to `get_registry_location()` method
* `taints` - added as a new feature for the `kubernetes-master` which applies taints this node by default to prevent scheduling


@Cynerva -- any issues on upgrade with the 
```configure_kubernetes_service('kubernetes-worker.prev_args.', ...'```
becomes
```configure_kubernetes_service('kubernetes-common.prev_args.', ...'```

this is the `key` field in some kind of key-value pair [here](https://github.com/charmed-kubernetes/layer-kubernetes-common/blob/53b511bb46c59c74a07265df497803978a54e662/lib/charms/layer/kubernetes_common.py#L456)

Reply:
It shouldn't be an issue. The prev_args handling comes from years ago when we used to configure the kubernetes services using individual snap config options, like so:

```
snap set kubelet kubeconfig=/root/cdk/kubeconfig network-plugin=cni ...
```

Snap config options are persistent, so we had to track and refer to prev_args to see if any arguments had been removed, so that we could explicitly set them to null to clear them [here](https://github.com/charmed-kubernetes/layer-kubernetes-common/blob/53b511bb46c59c74a07265df497803978a54e662/lib/charms/layer/kubernetes_common.py#L481-L483).

Since August 2019 ([PR](https://github.com/charmed-kubernetes/layer-kubernetes-common/pull/8)), we have used the `args` config option instead:

```
snap set kubelet args='--kubeconfig=/root/cdk/kubeconfig --network-plugin=cni ...'
```

One benefit of this approach is that, since we are overriding the entire `args` config every time we set it, we no longer need to explicitly clear previous args. We don't really need prev_args anymore, but it was left in to properly handle the transition from the old way to the new way.

Since that change, the prev_args value has always been set to `{}` [here](https://github.com/charmed-kubernetes/layer-kubernetes-common/blob/53b511bb46c59c74a07265df497803978a54e662/lib/charms/layer/kubernetes_common.py#L490-L492).

Indeed, if I check on a live 1.23 cluster running kubernetes-worker rev 838:

```
$ juju run --unit kubernetes-worker/0 -- source ../.venv/bin/activate '&&' python3 -c 'from charmhelpers.core.unitdata import kv; print(kv().get("kubernetes-worker.prev_args.kubelet"))'
{}
```

The only way I see this being an issue is if somebody upgrades directly from Charmed Kubernetes 1.15 to Charmed Kubernetes 1.24 without stepping through intermediate releases, a scenario which I would consider both extremely unsupported and extremely reckless.

Reply:
Having said all that, I could remove the prev_args key and related code entirely while I'm in here, if you prefer.

Reply:
i have no opinion on preference.  Just spotted it and thought it needed a comment